### PR TITLE
Support SARIF format result

### PR DESCRIPTION
**Pull request by shaopeng-gh**
Related issue: https://github.com/r-lib/lintr/issues/991
Usage: lintr::sarif_output(lintr::lint_dir("C:/repos/project1_root"), "sarif-results.sarif")

### CODE REVIEW

**Comment by AshesITR** [comment]
Hi @shaopeng-gh, thanks for the PR.
The pkgdown build failure can be fixed by adding `sarif_output` to `_pkgdown.yml`, e.g. on line 41.

File: R/lint.R [hunk 1]
```diff
@@ -561,6 +561,161 @@ checkstyle_output <- function(lints, filename = "lintr_results.xml") {
...
+      if (length(rule_index_exists) == 0L ||
+          is.na(rule_index_exists[1L])) {
+        rule_index_exists <- 0L
+      }
+    }
+
+    if (rule_index_exists == 0L) {
+      new_rule <- list(
+        id = lint$linter,
+        fullDescription = list(text = lint$message),
```
Comment (AshesITR) [inline R/lint.R:681]: Note that  `lint$message` is usually a customized message that incorporates context information for the specific piece of code that was linted, e.g. `undesirable_function_linter()` generates a message naming the concrete function that was detected. Is that okay for SARIF?

Comment (shaopeng-gh) [reply R/lint.R:681]: Yes, currently this field lint$message is used for both the each result's message and also rule description. Rule description is required field can not be empty. In each result's message user can see the correct message for that instance of the violation which is correct and what we want; the rule description should be static and not have any variable in the string. In the future if we have a separated property like lint$rule_description we should use it. Many the rules lint$message I see is static so this is a good starting version.

Comment (AshesITR) [reply R/lint.R:681]: Thanks for the explanation!

**Review by AshesITR — APPROVED** [review]

**Comment by shaopeng-gh** [comment]
@AshesITR thanks!
1. not familiar with the package upgrade process, can a new version to be published?
I will work on the starter workflow after a new version is published and available to use, I see sample in the internet:
`
- uses: r-lib/actions/setup-r-dependencies@v2
        with:
          extra-packages: lintr
`
This should be correct right? And should be good to get the latest version, after a new version is published.

2. the starter workflow will need a icon for the linter, I only find one at the top left of page https://github.com/r-lib, will use it, let me know if there is a better one. (And it can also be changed later)

**Comment by AshesITR** [comment]
Hi @shaopeng-gh, we just recently released version 3.0.0. Maybe we can get a patch release coming some time soon.
Regarding an icon, this is tracked by #981.

**Comment by shaopeng-gh** [comment]
@AshesITR 
sounds good, 
I have created the [lintr starter workflow pr](https://github.com/actions/starter-workflows/pull/1621) and set in draft status, 
we just need to replace the logo image when it is done in https://github.com/r-lib/lintr/issues/1188 (fyi @hadley), 
and set the pr to ready to review status after a new version released.

**Pull request merged**